testnet/stacks-node/src/event_dispatcher.rs

```diff
@@ -313,12 +311,112 @@ impl RewardSetEventPayload {
     }
 }
 
+fn send_request(
+    host: &str,
+    port: u16,
+    body: &[u8],
+    url: &Url,
+    timeout: Duration,
+) -> Result<String, std::io::Error> {
+    let addr = format!("{}:{}", host, port)
+        .to_socket_addrs()?
+        .next()
+        .ok_or_else(|| {
+            std::io::Error::new(std::io::ErrorKind::NotFound, "No valid address found")
+        })?;
+    let mut stream = TcpStream::connect_timeout(&addr, timeout)?;
+    stream.set_read_timeout(Some(timeout))?;
+    stream.set_write_timeout(Some(timeout))?;
+
+    let request = format!(
+        "POST {} HTTP/1.1\r\n\
+         Host: {}\r\n\
+         Content-Type: application/json\r\n\
+         Content-Length: {}\r\n\
+         Connection: close\r\n\
+         \r\n",
+        url.path(),
+        host,
+        body.len(),
+    );
+    debug!("Event dispatcher: Sending request"; "request" => &request);
+
+    stream.write_all(request.as_bytes())?;
+    stream.write_all(body)?;
+    stream.flush()?;
+    debug!("Event dispatcher: Request sent");
+
+    let mut response = Vec::new();
+    let mut buffer = [0; 512];
+    let mut headers_parsed = false;
+    let mut content_length = None;
+    let mut total_read = 0;
+
+    let start_time = Instant::now();
+
+    while total_read < content_length.unwrap_or(usize::MAX) {
```

Comment:
On a 64-bit system, `usize::MAX` is 2^64 - 1. Using `usize::MAX` here could allow a DoS attack, where the other side sends bytes until we're OOM and `response.extend_from_slice()` panics. We should set a limit for header length, returning an error if we haven't encountered `\r\n\r\n` after X bytes, and also a separate but larger limit for `content_len` that's less than `usize::MAX`

Reply:
Yeah, that's true. These observers should be trusted, since they have to be added to the node's config file, but it probably still makes sense to make this change.

Reply:
I think `u16::MAX` for header, and maybe `u32::MAX` for `content_len`. It's easily big enough for any legit response but I'd guess our VMs should have enough RAM (or at least swap space) that it shouldn't OOM

Reply:
Or maybe `1 << 30` for 1 GiB

Reply:
Your life here can be made easier by using our networking stack.

I haven't compiled this, but I'm happy to help you get it over the finish line if you want.

```rust
// TODO: get `conn_opts` from the node config file instead of the default
let conn_opts = ConnectionOptions::default();
// TODO: change `endpoint` to whatever the stacks observer endpoint is
let endpoint = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 30000);
let http = StacksHttp::new(endpoint, conn_opts);

let mut connection = NetworkConnection::new(http, &conn_opts, None);
let mut request_handle = connection.make_request_handle(0, timeout, 0);
request_handle.write_all(request.as_bytes())?;
request_handle.write_all(body)?;

debug!("Event dispatcher: Request sent");

// NOTE: your buffer should be page-aligned, and preferably multiple pages
let mut buffer = [0; 32768];
let start_time = Instant::now();
let mut flushed = false;
let mut response = None;

loop {
   if !flushed {
      flushed = request_handle.try_flush()?;
   }
   request_handle.send_data(&mut stream)?;
   let bytes_read = request_handle.recv_data(&mut stream)?;
   if bytes_read == 0 {
      break;
   }
   let mut msgs = request_handle.drain_inbox();
   if let Some(msg) = msgs.pop() {
      response = Some(msg);
      break;
   }
   if start_time.elapsed() >= timeout {
      return Err(std::io::Error::new(std::io::ErrorKind::TimedOut, "Response reading timed out"));
   }
}

// NOTE: this only succeeds if the content-type is text/plain.
// We also support application/json and application/octet-stream if you want.
let http_response : String = if let StacksHttpMessage::Response(resp) = response {
   resp
       .destruct()
       .1
       .try_into()
       .map_err(|e| std::io::Error::new(std::io::ErrorKind::Custom, "Did not receive text/plain data")))?;
}
else {
   return Err(std::io::Error::new(std::io::ErrorKind::Custom, "Did not receive an HTTP response"));
};

Ok(http_response)
```

let http_response_str : String = http_response.

Reply:
Ok, yeah I am having a little trouble with this. Did you intend for the calls to `recv_data` and `drain_inbox` to be on `connection` instead of `request_handle`?

Reply:
It seems like maybe I could just stay at the `NetworkReplyHandle` level, and call `write_all`, `flush`, then `recv`?

Reply:
Lemme take a crack

Reply:
Ok, I think these changes look good. Definitely could use some refactoring to make a simple http request like this much cleaner, but that can be done later.